subsys/net/lib/nrf_provisioning/src/nrf_provisioning_coap.c

```diff
@@ -0,0 +1,636 @@
+
+LOG_MODULE_REGISTER(nrf_provisioning_coap, CONFIG_NRF_PROVISIONING_LOG_LEVEL);
+
+#if !defined(CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG)
+#error "CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG not given"
+#endif
+
+#define TLS_SEC_TAG CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG
+#define COAP_HOST CONFIG_NRF_PROVISIONING_COAP_HOSTNAME
+#define PEER_PORT CONFIG_NRF_PROVISIONING_COAP_PORT
+
+#define MODEM_INFO_FWVER_SIZE 41
+#define CLIENT_VERSION "1"
+
+#define AUTH_MVER "mver=%s"
+#define AUTH_CVER "cver=%s"
+#define AUTH_PATH "p/auth"
+#define AUTH_API_TEMPLATE (AUTH_PATH "?" AUTH_MVER "&" AUTH_CVER)
+
+#define CMDS_PATH "p/cmd"
+#define CMDS_AFTER "after=%s"
+#define CMDS_MAX_RX_SZ "rxMaxSize=%s"
+#define CMDS_MAX_TX_SZ "txMaxSize=%s"
+#define CMDS_API_TEMPLATE (CMDS_PATH "?" CMDS_AFTER "&" CMDS_MAX_RX_SZ "&" CMDS_MAX_TX_SZ)
+
+static const char *resp_path = "p/rsp";
+
+static struct addrinfo *address;
+static struct coap_client client;
+
+static K_SEM_DEFINE(coap_response, 0, 1);
+
+int nrf_provisioning_coap_init(struct nrf_provisioning_mm_change *mmode)
+{
+	static bool initialized;
+	int ret;
+
+	nrf_provisioning_codec_init(mmode);
+
+	if (initialized) {
+		return 0;
+	}
+
+	/* Modem info library is used to obtain the modem FW version */
+	ret = modem_info_init();
+	if (ret) {
+		LOG_ERR("Modem info initialization failed, error: %d", ret);
+		return ret;
+	}
+
+	LOG_INF("Init CoAP client");
+	ret = coap_client_init(&client, NULL);
+	if (ret) {
+		LOG_ERR("Failed to initialize CoAP client, error: %d", ret);
+		return ret;
+	}
+
+	initialized = true;
+
+	return 0;
+}
+
+/* Setup TLS options on a given socket */
+static int tls_setup(int fd)
+{
+	int err;
+	int verify;
+	int session_cache;
+
+	/* Security tag that we have provisioned the certificate with */
+	const sec_tag_t tls_sec_tag[] = {
+		TLS_SEC_TAG,
+	};
+
+	LOG_INF("TLS setup");
+
+	/* Set up TLS peer verification */
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	verify = REQUIRED;
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, err %d", errno);
+		return err;
+	}
+
+	/* Associate the socket with the security tag
+	 * we have provisioned the certificate with.
+	 */
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_sec_tag, sizeof(tls_sec_tag));
+	if (err) {
+		LOG_ERR("Failed to setup TLS sec tag, err %d", errno);
+		return err;
+	}
+
+	if (IS_ENABLED(CONFIG_NRF_PROVISIONING_COAP_TLS_SESSION_CACHE)) {
+		session_cache = TLS_SESSION_CACHE_ENABLED;
+	} else {
+		session_cache = TLS_SESSION_CACHE_DISABLED;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SESSION_CACHE,
+					&session_cache, sizeof(session_cache));
+	if (err) {
+		LOG_ERR("Failed to set TLS_SESSION_CACHE option: %d", errno);
+		return err;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_HOSTNAME, COAP_HOST, strlen(COAP_HOST));
+	if (err) {
+		LOG_ERR("Failed to setup TLS hostname, err %d", errno);
+		return err;
+	}
+
```

Comment:
```suggestion
                       /* Enable connection ID */
                       uint32_t dtls_cid = TLS_DTLS_CID_ENABLED;

                       err = zsock_setsockopt(dl->fd, SOL_TLS, TLS_DTLS_CID, &dtls_cid,
                                              sizeof(dtls_cid));

```

Maybe control that by Kconfig, or document that we require Modem FW 1.3.5

Reply:
@eivindj-nordic suggested in my PR for the nrfcloud CoAP library that I not worry about Kconfigs or checking the modem firmware version, but just always try to set it, and if it fails with EOPNOTSUPP then you know CID is not available.